client/dashboard/sites/settings-repositories/use-install-github.ts

```diff
@@ -0,0 +1,152 @@
+import { useI18n } from '@wordpress/react-i18n';
+import { postLoginRequest } from '../../../state/login/utils';
+import { useAnalytics } from '../../app/analytics';
+
+const AUTHORIZATION_URL =
+	'https://public-api.wordpress.com/wpcom/v2/hosting/github/app-authorize?ux_mode=popup';
+const INSTALLATION_URL = 'https://public-api.wordpress.com/wpcom/v2/hosting/github/app-install';
+const NOTICE_ID = 'github-app-install-notice';
+
+interface OpenPopupOptions {
+	url: string;
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	onMessage( message: any, popup: Window ): void;
+}
+
+const openPopup = ( { url, onMessage }: OpenPopupOptions ) => {
+	let popup: Window | null;
+
+	try {
+		const width = 700;
+		const height = 600;
+
+		const top = window.screen.height / 2 - height / 2;
+		const left = window.screen.width / 2 - width / 2;
+
+		popup = window.open(
+			url,
+			undefined,
+			`popup=1,width=${ width },height=${ height },top=${ top },left=${ left }`
+		);
+	} catch {
+		return false;
+	}
+
+	const handleMessage = ( event: MessageEvent ) => {
+		const { data } = event;
+
+		if ( ! data || ! popup ) {
+			return;
+		}
+
+		onMessage( data, popup );
+	};
+
+	window.addEventListener( 'message', handleMessage );
+
+	const interval = window.setInterval( () => {
+		if ( ! popup || popup.closed ) {
+			clearInterval( interval );
+			window.removeEventListener( 'message', handleMessage );
+		}
+	}, 500 );
+
+	return true;
+};
+
+export const useInstallGithub = () => {
+	const { recordTracksEvent } = useAnalytics();
+	const { __ } = useI18n();
+	const queryClient = useQueryClient();
+	const { createInfoNotice, createErrorNotice } = useDispatch( noticesStore );
+	const { mutate: saveGitHubCredentials } = useMutation( saveGitHubCredentialsMutation() );
+
+	recordTracksEvent( 'calypso_hosting_github_app_open_auth_popup_requested' );
+
+	const authorizeApp = async ( { code }: { code: string } ) => {
+		const response = await postLoginRequest( 'exchange-social-auth-code', {
+			service: 'github',
+			auth_code: code,
+			client_id: config( 'wpcom_signup_id' ),
+			client_secret: config( 'wpcom_signup_key' ),
+		} );
+
+		if ( response.body.data.access_token ) {
+			await saveGitHubCredentials( { accessToken: response.body.data.access_token } );
+		}
+	};
+
+	const installGithub = ( {
+		onSuccess,
+		githubInstallationsError,
+	}: {
+		onSuccess: ( installationId: number ) => void;
+		githubInstallationsError: Error | null;
+	} ) => {
+		const openedPopup = openPopup( {
+			url:
+				githubInstallationsError?.name === 'UnauthorizedError'
+					? AUTHORIZATION_URL
+					: INSTALLATION_URL,
+			onMessage: async ( data, popup ) => {
+				if ( 'github-app-authorized' === data.type ) {
+					recordTracksEvent( 'calypso_hosting_github_app_authorised_success' );
+
+					try {
+						await authorizeApp( { code: data.code } );
+						queryClient.invalidateQueries( githubInstallationsQuery() );
```

Comment:
Can we move this invalidation to the `saveGitHubCredentials()` mutation's `onSuccess()` itself? The idea is invalidations should be part of the queries package, and components shouldn't need to know about it.

Reply:
I was thinking about it, but I consider `credentials` and `installations` as a not connected things, that's why decided to go this way, but I see you point, I will think about it and maybe go this way or propose anotehr option. Thanks, Fushar, for the initial review.

Reply:
I was thinking about it one more time - and it was great suggestion and totally makes sense to put it there, thanks oen more time ❤️